File: app/src/store/exportKind.ts
import type {
  LaserlineDeviceExport,
  PlanarIntrinsicsExport,
  RigExtrinsicsExport,
  RigHandeyeExport,
  RigHandeyeLaserlineExport,
  RigLaserlineDeviceExport,
  ScheimpflugIntrinsicsExport,
  SingleCamHandeyeExport,
} from "../types/generated/diagnose-wire";

/** Discriminated union over every calibration export the diagnose app can
 * load. Members are the generated (Rust-sourced) `*Export` interfaces, so
 * this stays in lockstep with the pipeline via `bun run generate:types`. */
export type CalibrationExport =
  | PlanarIntrinsicsExport
  | ScheimpflugIntrinsicsExport
  | SingleCamHandeyeExport
  | LaserlineDeviceExport
  | RigExtrinsicsExport
  | RigHandeyeExport
  | RigLaserlineDeviceExport
  | RigHandeyeLaserlineExport;

/** Discriminator for the loaded calibration export. */
export type ExportKind =
  | "planar_intrinsics"
  | "scheimpflug_intrinsics"
  | "single_cam_handeye"
  | "laserline_device"
  | "rig_extrinsics"
  | "rig_handeye"
  | "rig_handeye_laserline"
  | "rig_laserline_device"
  | "unknown";

// Distinguishing field names, pinned to the generated interfaces via
// `satisfies keyof …`. If a Rust field is renamed and the types are
// regenerated, these stop compiling — which is the whole point: the
// discriminator can no longer silently drift from the export shape.
const F_HANDEYE_MODE = "handeye_mode" satisfies keyof RigHandeyeExport &
  keyof SingleCamHandeyeExport &
  keyof RigHandeyeLaserlineExport;
const F_CAMERAS = "cameras" satisfies keyof RigExtrinsicsExport & keyof RigHandeyeExport;
const F_CAMERA = "camera" satisfies keyof SingleCamHandeyeExport;
const F_LASER_PLANES_RIG = "laser_planes_rig" satisfies keyof RigLaserlineDeviceExport &
  keyof RigHandeyeLaserlineExport;
const F_ESTIMATE = "estimate" satisfies keyof LaserlineDeviceExport;
const F_STATS = "stats" satisfies keyof LaserlineDeviceExport;
const F_PARAMS = "params" satisfies keyof PlanarIntrinsicsExport &
  keyof ScheimpflugIntrinsicsExport;

/** Narrows `value` to a plain (non-null, non-array) object — guards the
 * `params` cast below against malformed exports where the field exists
 * but isn't a record (e.g. a string or array). */
function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/** Sensor-model tag inside `*IntrinsicsExport.params.camera.sensor`, the only
 * place planar and Scheimpflug single-camera exports differ (their top-level
 * shapes are identical). */
function intrinsicsSensorTag(params: Record<string, unknown>): string | undefined {
  const camera = params.camera as { sensor?: { type?: unknown } } | undefined;
  const tag = camera?.sensor?.type;
  return typeof tag === "string" ? tag : undefined;
}

/** Classify a loaded export by its distinguishing required fields.
 *
 * Grounded in the generated `*Export` shapes rather than the JSON-sniffing
 * of the retired `exportShape.ts`: those probes read top-level `camera` /
 * `laser_planes_cam` fields that the real exports never carry (they live
 * under `params` / `estimate`). Probe order runs most-specific first —
 * laser-rig before rig, rig before single-camera. */
export function detectExportKind(data: unknown): ExportKind {
  if (data == null || typeof data !== "object") return "unknown";
  const d = data as Record<string, unknown>;
  const has = (k: string) => d[k] != null;

  // Rig + laser flavours carry per-camera laser planes in the rig frame.
  if (has(F_LASER_PLANES_RIG)) {
    return has(F_HANDEYE_MODE) ? "rig_handeye_laserline" : "rig_laserline_device";
  }
  // Single-camera laser device: bundle estimate + laser stats, no rig planes.
  if (has(F_ESTIMATE) && has(F_STATS)) return "laserline_device";
  // Multi-camera rig: a `cameras` array (with or without hand-eye).
  if (has(F_CAMERAS)) {
    return has(F_HANDEYE_MODE) ? "rig_handeye" : "rig_extrinsics";
  }
  // Single-camera hand-eye: one `camera` plus a hand-eye mode.
  if (has(F_CAMERA) && has(F_HANDEYE_MODE)) return "single_cam_handeye";
  // Single-camera intrinsics: `params`; planar vs Scheimpflug by sensor model.
  // `params` must be a plain object — a malformed export where it's some
  // other JSON type (string, array, …) falls through to "unknown" instead
  // of silently defaulting to planar.
  const params = d[F_PARAMS];
  if (isPlainObject(params)) {
    return intrinsicsSensorTag(params) === "scheimpflug"
      ? "scheimpflug_intrinsics"
      : "planar_intrinsics";
  }
  return "unknown";
}

/** Human-readable label for an export kind, for the diagnose header. */
export function exportKindLabel(kind: ExportKind): string {
  switch (kind) {
    case "planar_intrinsics":
      return "Planar intrinsics";
    case "scheimpflug_intrinsics":
      return "Scheimpflug intrinsics";
    case "single_cam_handeye":
      return "Single-cam hand-eye";
    case "laserline_device":
      return "Laserline device";
    case "rig_extrinsics":
      return "Rig extrinsics";
    case "rig_handeye":
      return "Rig hand-eye";
    case "rig_handeye_laserline":
      return "Rig hand-eye + laserline";
    case "rig_laserline_device":
      return "Rig + laserline";
    case "unknown":
      return "Unknown export";
  }
}

File: crates/vision-calibration-pipeline/src/common/export_kind.rs
//! Export-type discriminator (R7).
//!
//! Every pipeline `*Export` carries a `kind: ExportKind` tag as its first
//! field so consumers narrow on a single value instead of probing which
//! required fields are present. The tag serializes snake_case
//! (`"planar_intrinsics"`, …) — the same vocabulary as the problem-module
//! names — and is **required** on deserialize: a missing tag is an error, not
//! a silent default. The 0.7.0 breaking window regenerates every committed
//! export so nothing relies on the pre-tag shape.

use serde::{Deserialize, Serialize};

/// Identifies which calibration `*Export` a JSON payload holds.
///
/// One variant per pipeline problem type, serialized as the snake_case module
/// name in the `kind` field of every `*Export`. Consumers (the desktop app's
/// `detectExportKind`, downstream tooling) read this single tag rather than
/// sniffing field presence.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
#[serde(rename_all = "snake_case")]
pub enum ExportKind {
    /// [`PlanarIntrinsicsExport`](crate::planar_intrinsics::PlanarIntrinsicsExport).
    PlanarIntrinsics,
    /// [`ScheimpflugIntrinsicsExport`](crate::scheimpflug_intrinsics::ScheimpflugIntrinsicsExport).
    ScheimpflugIntrinsics,
    /// [`SingleCamHandeyeExport`](crate::single_cam_handeye::SingleCamHandeyeExport).
    SingleCamHandeye,
    /// [`LaserlineDeviceExport`](crate::laserline_device::LaserlineDeviceExport).
    LaserlineDevice,
    /// [`RigExtrinsicsExport`](crate::rig_extrinsics::RigExtrinsicsExport).
    RigExtrinsics,
    /// [`RigHandeyeExport`](crate::rig_handeye::RigHandeyeExport).
    RigHandeye,
    /// [`RigLaserlineDeviceExport`](crate::rig_laserline_device::RigLaserlineDeviceExport).
    RigLaserlineDevice,
    /// [`RigHandeyeLaserlineExport`](crate::rig_handeye_laserline::RigHandeyeLaserlineExport).
    RigHandeyeLaserline,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn serializes_snake_case() {
        assert_eq!(
            serde_json::to_string(&ExportKind::PlanarIntrinsics).unwrap(),
            "\"planar_intrinsics\""
        );
        assert_eq!(
            serde_json::to_string(&ExportKind::RigHandeyeLaserline).unwrap(),
            "\"rig_handeye_laserline\""
        );
    }

    #[test]
    fn roundtrips_every_variant() {
        for kind in [
            ExportKind::PlanarIntrinsics,
            ExportKind::ScheimpflugIntrinsics,
            ExportKind::SingleCamHandeye,
            ExportKind::LaserlineDevice,
            ExportKind::RigExtrinsics,
            ExportKind::RigHandeye,
            ExportKind::RigLaserlineDevice,
            ExportKind::RigHandeyeLaserline,
        ] {
            let json = serde_json::to_string(&kind).unwrap();
            let back: ExportKind = serde_json::from_str(&json).unwrap();
            assert_eq!(kind, back);
        }
    }
}
